File: shared/hooks/lib/json_tool.py
#!/usr/bin/env python3
"""Hook JSON codec. CPython stdlib only. Invoked by lib/json.sh."""
import json
import os
import sys


def walk(data, path):
    cur = data
    for part in path.split("."):
        if cur is None:
            return None
        if isinstance(cur, list) and part.isdigit():
            i = int(part)
            if i < 0 or i >= len(cur):
                return None
            cur = cur[i]
        elif isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur


def first(data, paths):
    for path in paths:
        value = walk(data, path)
        if value is not None:
            return value
    return None


def jtype(value):
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, (int, float)):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "object"
    return "string"


def as_text(value):
    if value is None:
        return ""
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return value
    if isinstance(value, (int, float)):
        return json.dumps(value)
    return json.dumps(value, ensure_ascii=False)


def posix_sq(value):
    text = as_text(value)
    return "'" + text.replace("'", "'\\''") + "'"


def load_stdin():
    raw = sys.stdin.read()
    if raw.startswith("\ufeff"):
        raw = raw[1:]
    raw = raw.replace("\r\n", "\n").replace("\r", "")
    if not raw.strip():
        raise ValueError("empty")
    return json.loads(raw)


def emit_obj(obj):
    sys.stdout.write(json.dumps(obj, separators=(",", ":"), ensure_ascii=False))
    sys.stdout.write("\n")


def env(name):
    return os.environ.get(name, "")


def cmd_emit(kind):
    msg = env("KLEOS_JSON_MSG")
    reason = env("KLEOS_JSON_REASON") or "deny"
    agent = env("KLEOS_JSON_AGENT")
    if kind == "allow":
        emit_obj({"permission": "allow", "agent_message": msg} if msg else {"permission": "allow"})
        return 0
    if kind == "deny":
        obj = {"permission": "deny", "user_message": msg, "reason": reason}
        if agent:
            obj["agent_message"] = agent
        emit_obj(obj)
        return 0
    if kind == "ask":
        obj = {"permission": "ask", "user_message": msg, "reason": reason}
        if agent:
            obj["agent_message"] = agent
        emit_obj(obj)
        return 0
    if kind == "continue":
        cont = env("KLEOS_JSON_CONTINUE") != "false"
        if not cont:
            obj = {"continue": False, "reason": reason or "block"}
            if msg:
                obj["user_message"] = msg
            emit_obj(obj)
            return 0
        emit_obj({"continue": True, "user_message": msg} if msg else {"continue": True})
        return 0
    return 1


def cmd_decode_submit():
    data = load_stdin()
    prompt = first(data, ("prompt", "user_prompt", "message", "text"))
    sys.stdout.write("PROMPT=%s\n" % posix_sq(prompt if isinstance(prompt, str) else as_text(prompt)))
    return 0


def cmd_decode_shell():
    data = load_stdin()
    cmd = first(data, ("command", "tool_input.command", "tool_input.cmd"))
    cwd = first(data, ("cwd", "tool_input.cwd", "workspace_roots.0"))
    sys.stdout.write("TYPE=%s\n" % posix_sq(jtype(cmd)))
    sys.stdout.write("CMD=%s\n" % posix_sq(cmd if isinstance(cmd, str) else as_text(cmd) if cmd is not None else ""))
    sys.stdout.write("CWD=%s\n" % posix_sq(cwd if isinstance(cwd, str) else as_text(cwd)))
    return 0


def cmd_decode_read():
    data = load_stdin()
    path = first(data, ("file_path", "path", "tool_input.file_path", "tool_input.path"))
    sys.stdout.write("FILE_PATH=%s\n" % posix_sq(path if isinstance(path, str) else as_text(path)))
    return 0


def cmd_file_valid(path):
    with open(path, "r") as handle:
        json.load(handle)
    return 0


def cmd_has_scripts_test(path):
    with open(path, "r") as handle:
        data = json.load(handle)
    scripts = data.get("scripts") if isinstance(data, dict) else None
    if isinstance(scripts, dict) and scripts.get("test"):
        return 0
    return 1


def main(argv):
    if len(argv) < 2:
        return 1
    cmd = argv[1]
    try:
        if cmd == "ping":
            sys.stdout.write("ok\n")
            return 0
        if cmd == "valid":
            load_stdin()
            return 0
        if cmd == "emit":
            if len(argv) < 3:
                return 1
            return cmd_emit(argv[2])
        if cmd == "decode-submit":
            return cmd_decode_submit()
        if cmd == "decode-shell":
            return cmd_decode_shell()
        if cmd == "decode-read":
            return cmd_decode_read()
        if cmd == "file-valid":
            return cmd_file_valid(argv[2])
        if cmd == "has-scripts-test":
            return cmd_has_scripts_test(argv[2])
        return 1
    except (ValueError, TypeError, json.JSONDecodeError, OSError, IndexError):
        return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: shared/hooks/lib/json_tool.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");

function walk(data, path) {
  let cur = data;
  for (const part of String(path).split(".")) {
    if (cur == null) return undefined;
    if (Array.isArray(cur) && /^\d+$/.test(part)) {
      const i = Number(part);
      cur = i >= 0 && i < cur.length ? cur[i] : undefined;
    } else if (typeof cur === "object") {
      cur = cur[part];
    } else {
      return undefined;
    }
  }
  return cur;
}

function first(data, paths) {
  for (const path of paths) {
    const value = walk(data, path);
    if (value !== undefined && value !== null) return value;
  }
  return undefined;
}

function jtype(value) {
  if (value === null || value === undefined) return "null";
  if (typeof value === "boolean") return "boolean";
  if (typeof value === "number") return "number";
  if (typeof value === "string") return "string";
  if (Array.isArray(value)) return "array";
  if (typeof value === "object") return "object";
  return "string";
}

function asText(value) {
  if (value === undefined || value === null) return "";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return value;
  return JSON.stringify(value);
}

function posixSq(value) {
  return "'" + asText(value).replace(/'/g, "'\\''") + "'";
}

function loadStdin() {
  let raw = fs.readFileSync(0, "utf8");
  if (raw.charCodeAt(0) === 0xfeff) raw = raw.slice(1);
  raw = raw.replace(/\r\n/g, "\n").replace(/\r/g, "");
  if (!raw.trim()) throw new Error("empty");
  return JSON.parse(raw);
}

function emitObj(obj) {
  process.stdout.write(JSON.stringify(obj) + "\n");
}

function env(name) {
  return process.env[name] || "";
}

function cmdEmit(kind) {
  const msg = env("KLEOS_JSON_MSG");
  const reason = env("KLEOS_JSON_REASON") || "deny";
  const agent = env("KLEOS_JSON_AGENT");
  if (kind === "allow") {
    emitObj(msg ? { permission: "allow", agent_message: msg } : { permission: "allow" });
    return 0;
  }
  if (kind === "deny" || kind === "ask") {
    const obj = { permission: kind, user_message: msg, reason };
    if (agent) obj.agent_message = agent;
    emitObj(obj);
    return 0;
  }
  if (kind === "continue") {
    const cont = env("KLEOS_JSON_CONTINUE") !== "false";
    if (!cont) {
      const obj = { continue: false, reason: reason || "block" };
      if (msg) obj.user_message = msg;
      emitObj(obj);
      return 0;
    }
    emitObj(msg ? { continue: true, user_message: msg } : { continue: true });
    return 0;
  }
  return 1;
}

function main(argv) {
  const cmd = argv[2];
  if (!cmd) return 1;
  if (cmd === "ping") {
    process.stdout.write("ok\n");
    return 0;
  }
  if (cmd === "valid") {
    loadStdin();
    return 0;
  }
  if (cmd === "emit") return cmdEmit(argv[3]);
  if (cmd === "decode-submit") {
    const data = loadStdin();
    const prompt = first(data, ["prompt", "user_prompt", "message", "text"]);
    process.stdout.write("PROMPT=" + posixSq(typeof prompt === "string" ? prompt : asText(prompt)) + "\n");
    return 0;
  }
  if (cmd === "decode-shell") {
    const data = loadStdin();
    const c = first(data, ["command", "tool_input.command", "tool_input.cmd"]);
    const cwd = first(data, ["cwd", "tool_input.cwd", "workspace_roots.0"]);
    process.stdout.write("TYPE=" + posixSq(jtype(c === undefined ? null : c)) + "\n");
    process.stdout.write(
      "CMD=" + posixSq(typeof c === "string" ? c : c == null ? "" : asText(c)) + "\n"
    );
    process.stdout.write("CWD=" + posixSq(typeof cwd === "string" ? cwd : asText(cwd)) + "\n");
    return 0;
  }
  if (cmd === "decode-read") {
    const data = loadStdin();
    const p = first(data, ["file_path", "path", "tool_input.file_path", "tool_input.path"]);
    process.stdout.write("FILE_PATH=" + posixSq(typeof p === "string" ? p : asText(p)) + "\n");
    return 0;
  }
  if (cmd === "file-valid") {
    JSON.parse(fs.readFileSync(argv[3], "utf8"));
    return 0;
  }
  if (cmd === "has-scripts-test") {
    const data = JSON.parse(fs.readFileSync(argv[3], "utf8"));
    return data && data.scripts && data.scripts.test ? 0 : 1;
  }
  return 1;
}

try {
  process.exit(main(process.argv));
} catch (err) {
  process.exit(2);
}
